XD-3622: File Source

garyrussell opened the pull request:
JIRA: https://jira.spring.io/browse/XD-3622

Port file source from XD; add regex for pattern.

garyrussell committed:
XD-3622: File Source

JIRA: https://jira.spring.io/browse/XD-3622

Port file source from XD; add regex for pattern.

artembilan commented on file-source/README.adoc:
```diff
@@ -0,0 +1,33 @@
+= Spring Cloud Stream File Sink
```
??? Is something wrong here?
I wonder if we should tell here about `File Source` instead...

garyrussell replied:
Strange - I remember changing it but somehow the changes are lost.

garyrussell replied:
Found it; also the parent pom change etc; I did a `git add .` from within the module.

artembilan commented on file-source/src/main/java/org/springframework/cloud/stream/module/file/source/FileSource.java:
```diff
@@ -0,0 +1,128 @@
+
+/**
+ * Creates a {@link FileReadingMessageSource} bean and registers it as a
+ * Inbound Channel Adapter that sends messages from the Source output channel.
```
... sends messages s/from/to the Source output channel.

ericbottard replied:
That sentence look awkward to me as well, but I think I understand Gary's "point of view" here, namely looking at the module from its exterior.

ericbottard commented:
Don't forget to add this new module to the top-level `pom.xml` :)

ericbottard commented on file-source/pom.xml:
```diff
@@ -0,0 +1,34 @@
+	</parent>
+
+	<properties>
+		<start-class>org.springframework.cloud.stream.module.file.FileSourceApplication</start-class>
```
wrong package

markfisher replied:
file-sink's pom will need to be updated also

ericbottard commented on file-source/src/main/java/org/springframework/cloud/stream/module/file/source/FileSourceProperties.java:
```diff
@@ -0,0 +1,93 @@
+	/**
+	 * The directory to poll for new files.
+	 */
+	private String directory = DEFAULT_DIR;
```
Wondering if this would make sense to turn this property to a `java.io.File` directly

ericbottard commented on file-source/src/main/java/org/springframework/cloud/stream/module/file/source/FileSourceProperties.java:
```diff
@@ -0,0 +1,93 @@
+	private String directory = DEFAULT_DIR;
+
+	/**
+	 * Set to true to include an {@code AcceptOnceFileListFilter}.
```
Will need to have a look at what the usage of `{@code ..}` implies in generated json metadata.

garyrussell replied:
Shall I remove it?

ericbottard replied:
I will have a look at what is generated (you can too in `target/classes` after a maven build)

garyrussell replied:
Looks like we need to remove it...

```
      "description": "Set to true to include an {@code AcceptOnceFileListFilter}.",
```

ericbottard commented on file-source/src/main/java/org/springframework/cloud/stream/module/file/source/FileSourceProperties.java:
```diff
@@ -0,0 +1,93 @@
+	/**
+	 * A regex pattern to match files.
+	 */
+	private String filenameRegex;
```
Could turn this to a `Pattern` (with appropriate Converter) to provide earlier binding error feedback

garyrussell replied:
Doh; yes; I did that in (S)FTP after your suggestion.

ericbottard commented on file-source/src/test/java/org/springframework/cloud/stream/module/file/source/FileSourceTests.java:
```diff
@@ -0,0 +1,196 @@
+	private static final String ROOT_DIR = TMPDIR + File.separator + "dataflow-tests"
+			+ File.separator + "input";
+
+	protected File atomicFileCreate(String filename) throws FileNotFoundException, IOException {
```
I don't understand why, if this method takes a parameter, it should be the name of the "hidden" file.

garyrussell replied:
Not sure what you mean; it's called to create `test.txt` and `test.foo` (which should not be found by the pattern tests).

ericbottard replied:
What I'm saying is that (unless I'm mistaken), the file the caller gets in return is always named `test.txt` (from line 60).
The intermediate file, however, gets a name from the parameter. But we don't really care about that, as it is transient.

ericbottard commented on file-source/src/test/java/org/springframework/cloud/stream/module/file/source/FileSourceTests.java:
```diff
@@ -0,0 +1,196 @@
+	@Autowired
+	protected MessageCollector messageCollector;
+
+	@IntegrationTest({ "directory = ${java.io.tmpdir}${file.separator}dataflow-tests${file.separator}input",
```
Look at this @markfisher . Gary is full of resources! I believe we can change some of out IT to use this approach

markfisher replied:
Eh? I'm pretty sure he copied that idea from my `FileSinkTests`, e.g. https://github.com/spring-cloud/spring-cloud-stream-modules/blob/master/file-sink/src/test/java/org/springframework/cloud/stream/module/file/FileSinkTests.java#L54

ericbottard replied:
Hmm, did not see that. Let's not get into a paternity war, shall we ;) Anyway, this is good (new) news to me!

markfisher replied:
Well, of course I also copied it, so all credit to @trisberg (or whomever he copied it from):
https://github.com/spring-cloud/spring-cloud-stream-modules/blob/master/hdfs-sink/src/test/java/org/springframework/cloud/stream/module/hdfs/sink/HdfsSinkIntegrationTests.java#L54

...and don't get me wrong, I agree Gary is full of resources ;)

ericbottard commented on file-source/src/test/java/org/springframework/cloud/stream/module/file/source/FileSourceTests.java:
```diff
@@ -0,0 +1,196 @@
+			String filename = "test.txt";
+			File file = atomicFileCreate(filename);
+			Message<?> received = messageCollector.forChannel(source.output()).poll(10, TimeUnit.SECONDS);
+			assertNotNull(received);
```
I think the returned **message** can never be `null` (same can't be said of payload variant methods)

ericbottard commented on file-source/src/test/java/org/springframework/cloud/stream/module/file/source/FileSourceTests.java:
```diff
@@ -0,0 +1,196 @@
+			Message<?> received = messageCollector.forChannel(source.output()).poll(10, TimeUnit.SECONDS);
+			assertNotNull(received);
+			assertEquals(file, received.getPayload());
+			received = messageCollector.forChannel(source.output()).poll(300, TimeUnit.MILLISECONDS);
```
Apparently I was wrong above. I thought timeout threw an exception. nvm

artembilan replied:
Yes, you were: the `poll()` works the same way as it is in the `java.util.concurrent.BlockingQueue` impl

artembilan commented on file-source/src/main/java/org/springframework/cloud/stream/module/file/source/FileSource.java:
```diff
@@ -0,0 +1,128 @@
+ * @author Gary Russell
+ */
+@EnableBinding(Source.class)
+@Import({PeriodicTriggerConfiguration.class, SpelExpressionConverterConfiguration.class})
```
I wonder if `SpelExpressionConverterConfiguration` can be declared implicitly and always unconditionally.
From one side it just adds an extra `Converter`. From another side it doesn't do anything with this `FileSource` `@Configuration`.

ericbottard replied:
Yes, we thought about that with @markfisher. I'd vote for it too, maybe making it conditional on the absence of a converter for that already, for the rare case where the user would want his own.
Same for other converters that I think we could/should provide

artembilan commented on file-source/src/main/java/org/springframework/cloud/stream/module/file/source/FileSource.java:
```diff
@@ -0,0 +1,128 @@
+
+	@Bean
+	public PollerMetadata poller() {
+		return Pollers.trigger(trigger).maxMessagesPerPoll(this.maxMessagesProperties.getMaxMessages()).get();
```
Looks like this is redundant and can be as a part of `PeriodicTriggerConfiguration`.
As well as `MaxMessagesProperties` should be moved there.

artembilan commented on file-source/src/main/java/org/springframework/cloud/stream/module/file/source/FileSource.java:
```diff
@@ -0,0 +1,128 @@
+	MaxMessagesProperties maxMessagesProperties;
+
+	@Autowired
+	Trigger trigger;
```
Is this fine that they are without `private`, not like a couple above ?

artembilan commented on file-source/src/main/java/org/springframework/cloud/stream/module/file/source/FileSource.java:
```diff
@@ -0,0 +1,128 @@
+	}
+
+	@Autowired
+	Source source;
```
Can some one explain me why do we need such a simple interface injection, if it doesn't have so much power only `output()`? 
Why just don't inject that `MessageChannel output` ?
Thanks

artembilan commented on file-source/src/main/java/org/springframework/cloud/stream/module/file/source/FileSource.java:
```diff
@@ -0,0 +1,128 @@
+		case contents:
+			flowBuilder.enrichHeaders(Collections.<String, Object>singletonMap(MessageHeaders.CONTENT_TYPE,
+					"application/octet-stream"))
+					.transform(new FileToByteArrayTransformer());
```
I wonder if we should support `CONTENT_TYPE` header in those `AbstractFilePayloadTransformer`s directly...
Like we do with the `ObjectToJsonTransformer`.

artembilan replied:
`Transformers.fileToByteArray()` BTW

garyrussell replied:
Please create a 4.3 JIRA for the CT.

artembilan commented on file-source/src/main/java/org/springframework/cloud/stream/module/file/source/FileSource.java:
```diff
@@ -0,0 +1,128 @@
+
+	private IntegrationFlowBuilder enhanceFlowForReadingMode(IntegrationFlowBuilder flowBuilder) {
+		switch (this.fileConsumerProperties.getMode()) {
+		case contents:
```
Interesting, why enum const is in lower case...

ericbottard replied:
because it's prettier IMO :sparkling_heart: 
Note that at time of binding, the Converter accepts any case

artembilan commented on file-source/src/main/java/org/springframework/cloud/stream/module/file/source/FileSource.java:
```diff
@@ -0,0 +1,128 @@
+		case lines:
+			flowBuilder.enrichHeaders(Collections.<String, Object>singletonMap(MessageHeaders.CONTENT_TYPE,
+					"text/plain"))
+					.split(new FileSplitter(true, this.fileConsumerProperties.getWithMarkers()), null);
```
`.split(Files.splitter(true, this.fileConsumerProperties.getWithMarkers()))`
Not sure what the problem you have here with the second argument, but that works fine even with the `new FileSplitter()`

Minor, of course...

garyrussell replied:
True, with 1.1, but we're using 1.0.1 for some reason.

ericbottard commented on file-source/src/main/java/org/springframework/cloud/stream/module/file/source/FileSource.java:
```diff
@@ -0,0 +1,128 @@
+			break;
+		case lines:
+			flowBuilder.enrichHeaders(Collections.<String, Object>singletonMap(MessageHeaders.CONTENT_TYPE,
+					"text/plain"))
```
Can we really assume text**/plain** here? As opposed to `text/?` ie just `text`

artembilan replied:
That is what the next `FileSplitter` produces: only File lines as strings

ericbottard replied:
What I'm saying is: we don't know that it is `plain`. Just because we split on newlines, doesn't mean that it's not, say, `text/xml`. Not a big deal, as we can later force the output type. But may be too big of an assumption

garyrussell replied:
This was a straight port from XD which also uses `text/plain`

garyrussell committed:
XD-3622: Polishing; PR Comments + Missing Files

garyrussell commented:
Pushed

garyrussell committed:
XD-3622: Refactor File/(S)FTP Common Code

garyrussell commented:
Pushed another commit to factor out common code from File/(S)FTP Sources.

Bumped DSL version to 1.1.0.

garyrussell committed:
Remove {@Code ...} from Javadocs

garyrussell committed:
Fix Hidden File Tests

The hidden files were not actually created so the tests actually didn't test
anything.

garyrussell commented:
Doh sorry - it's supposed to create a hidden file; bad refactoring; pushed.

ericbottard commented:
Merged as c378ca2

Also added https://github.com/spring-cloud/spring-cloud-dataflow/commit/3ec711445dd382fef988d28135c2db74f711b4a9
